services/skus/service.go

```diff
@@ -2282,6 +2292,122 @@ func (s *Service) processSubmitReceipt(ctx context.Context, req model.ReceiptReq
 	return rcpt, nil
 }
 
+const errUnknownAction = model.Error("skus: unknown action") // TODO better message and rename
+
+func (s *Service) processRadomEvent(ctx context.Context, event radom.Event) error {
+	if !event.ShouldProcess() {
+		return nil
+	}
+
+	tx, err := s.Datastore.RawDB().BeginTxx(ctx, nil) // TODO
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	if err := s.processRadomEventTx(ctx, tx, event); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func (s *Service) processRadomEventTx(ctx context.Context, dbi sqlx.ExtContext, event radom.Event) error {
+	switch {
+	case event.IsNewSub():
+		oid, err := event.OrderID()
+		if err != nil {
+			return err
+		}
+
+		subID, err := event.SubID()
+		if err != nil {
+			return err
+		}
+
+		rsub, err := s.radomClient.GetSubscription(ctx, subID)
+		if err != nil {
+			return err
+		}
+
+		nxtB, err := rsub.NextBillingDate()
+		if err != nil {
+			return err
+		}
+
+		expAt := nxtB.Add(24 * time.Hour)
+
+		paidAt, err := rsub.LastPaid()
+		if err != nil {
+			return err
+		}
+
+		if err := s.renewOrderWithExpPaidTimeTx(ctx, dbi, oid, expAt, paidAt); err != nil {
+			return err
+		}
+
+		if err := s.orderRepo.AppendMetadata(ctx, dbi, oid, "externalID", subID.String()); err != nil {
+			return err
+		}
+
+		if err := s.orderRepo.AppendMetadata(ctx, dbi, oid, "paymentProcessor", model.RadomPaymentMethod); err != nil {
+			return err
+		}
+
+	case event.ShouldRenew():
+		subID, err := event.SubID()
+		if err != nil {
+			return err
+		}
+
+		ord, err := s.orderRepo.GetByExternalID(ctx, dbi, subID.String())
+		if err != nil {
+			return err
+		}
+
+		rsub, err := s.radomClient.GetSubscription(ctx, subID)
+		if err != nil {
+			return err
+		}
+
+		nxtB, err := rsub.NextBillingDate()
+		if err != nil {
+			return err
+		}
+
+		expAt := nxtB.Add(24 * time.Hour)
+
+		paidAt, err := rsub.LastPaid()
+		if err != nil {
+			return err
+		}
+
+		if err := s.renewOrderWithExpPaidTimeTx(ctx, dbi, ord.ID, expAt, paidAt); err != nil {
+			return err
+		}
+
+	case event.ShouldCancel():
```

Comment:
Once a user cancels do we ever revisit their data stored and remove it? Just trying to figure out what our retention is here.

Reply:
AFAIK we do not delete the data once a subscription is cancelled.

Reply:
Ok, this is probably something we should follow up on, but let's handle it separately.